### PR TITLE
#360 - New document fixtures from real data

### PR DESCRIPTION
**Pull request by blms**
## What this PR does
- Adds `generate_fixtures` management command to generate document test fixtures
- Adds new document fixtures generated with the following command:
   ```sh
   python manage.py generate_fixtures 444 912 2142 2532 2811 2969 3497 3504 3879 4744 5626 6343 19612 19633 23342 29181 29324 30176 30208 30661 30818 30820 30824 30825 30833 30851 31094 31134 31950 32012 32644 34177 34373
   ```
- Uses those fixtures in visual tests (Percy and Lighthouse)
- Adds natural key support to `DocumentType` model

<details><summary>Previous notes</summary>

## General notes
- So far I have generated the fixtures, and tested that I can load, index, and display them using a temp database.
- To get it to work, I used `call_command("loaddata", "test_documents.json")` in `create_test_site.py` right before `make_fragment()` is called. But this might not be the best place for it and I haven't committed it on this branch yet—see questions below.

## Questions
- Since it was pretty complex to generate fixtures, as you can see from the steps below, I wonder where it makes the most sense to document this?
- Since my data migration from previous work created some `DocumentType`s, the ones used for testing needed to avoid PK and name collisions. Because of that, I manually removed the colliding document types from the fixture JSON. For the two remaining document types, I just incremented their PKs. Probably not the best move because now none of the document types match the real data—the PKs are all shuffled—but not sure how else to handle. Any suggestions?
  - Maybe calling `loaddata` first before `migrate`? In that case, `loaddata` would need to be given its own step in `visual_tests.yml`, before `migrate`.
  - It also might not matter that these are mismatched, but it slightly defeats the purpose of using real data.
- Should I replace the existing `make_fragment` ,`make_multifragment`, `make_document`, `make_join` calls in `create_test_site` with my new fixtures? Might need to then fix a bunch of failing tests to match the new fixtures, unless the data in those old methods are definitely going to match the new fixtures already.

## Steps to generate fixtures
1. I installed `django-fixture-magic` and added it to `requirements/dev.txt` and `geniza/settings/environments/development.py`.
2. I ran the normal documents dump with some selected PKs that I think represent a good range (including some weird outliers, so that we make sure to handle them gracefully in the UI).
    ```sh
    python manage.py dump_object corpus.document --query '{"pk__in": [444, 912, 2142, 2532, 2811, 2969, 3497, 3504, 3879, 4744, 5626, 6343, 19612, 19633, 23342, 29181, 29324, 30176, 30208, 30661, 30818, 30820, 30824, 30825, 30833, 30851, 31094, 31134, 31950, 32012, 32644, 34177, 34373]}' > ./geniza/corpus/fixtures/documents.json
    ```
3. I had to create a custom dump, and use it in addition to the normal documents dump, in order to get (almost) all the relations represented in the dumps. You can see that custom dump in `geniza/settings/environments/development.py`.
    ```sh
    python manage.py custom_dump document 444 912 2142 2532 2811 2969 3497 3504 3879 4744 5626 6343 19612 19633 23342 29181 29324 30176 30208 30661 30818 30820 30824 30825 30833 30851 31094 31134 31950 32012 32644 34177 34373 > ./geniza/corpus/fixtures/document_dependents.json
    ```
4. In _addition_ to that custom dump, I had to do a manual dump and filter of `Authorships`. I guess `django-fixture-magic` doesn't like many-to-many-through relations. To do that I did the following:
    ```sh
    python manage.py dumpdata footnotes.authorship > ./geniza/corpus/fixtures/authorships.json
    ```
5. I wrote and ran the following script to filter out the unneeded authorships:
    ```python
    import json
    from geniza.corpus.models import Document
    
    pk_list = [444, 912, 2142, 2532, 2811, 2969, 3497, 3504, 3879, 4744, 5626, 6343, 19612, 19633, 23342, 29181, 29324, 30176, 30208, 30661, 30818, 30820, 30824, 30825, 30833, 30851, 31094, 31134, 31950, 32012, 32644, 34177, 34373]
    qs = Document.objects.filter(pk__in=pk_list)
    source_ids = list(set(qs.values_list("footnotes__source", flat=True)))
    source_ids = [x for x in source_ids if x is not None]
    authorships = []
    
    with open("geniza/corpus/fixtures/authorships.json") as fixture:
        all_authorships = json.load(fixture)
    
    for authorship in all_authorships:
        if authorship["fields"]["source"] in source_ids:
            authorships.append(authorship)
    
    with open("geniza/corpus/fixtures/authorships.json", "w") as new_fixture:
        json.dump(authorships, new_fixture)
    ```
6. I then ran the following to merge and de-dupe the fixtures:
    ```sh
    python manage.py merge_fixtures ./geniza/corpus/fixtures/documents.json ./geniza/corpus/fixtures/document_dependents.json ./geniza/corpus/fixtures/authorships.json  > ./geniza/corpus/fixtures/test_documents.json
    ```
7. Finally, I deleted the unmerged fixtures, and manually edited the merged fixture to deal with the `DocumentType` collisions as mentioned above.

</details>

### CODE REVIEW

**Review by rlskoeser — COMMENTED** [review]
Wow, this is a lot! Thanks for investigating and figuring all this out. I've never actually defined a custom dump; not sure how I didn't know it didn't handle through tables.

To get the authorship records, I think we should be able to use the query parameter. Does this get the right set?
```
./manage.py dump_object footnotes.authorship --query '{"source__footnote__object_id__in": [444, 912]}'
```
For documentation: let's start a new `DEVELOPERNOTES` document at the top level parallel to the readme. (Example in a different project: https://github.com/Princeton-CDH/ppa-django/blob/main/DEVELOPERNOTES.rst )  Probably some of what's in the readme now could shift over, but don't worry about that. You can just start the document and add section about this.

Alternately: I wonder if it's worth making a minimal manage command to generate this fixture? Define the list of ids once and then it runs dump and merge commands? 

Regarding the id collision: we should be using natural keys for the export, which we need to specify when we call `dump_object`. We've defined natural keys this for some of our models but not all; at the very least, based on your notes I think we need to  add one for `DocumentType`. We could add for others — I can't tell from your notes how many others are needed; let's focus on the ones we need to get this fixture where we need it to proceed on the other development work.

My inclination is to drop the test fixtures from our test site and just use the new fixture for both lighthouse and percy — then the test fixtures will be exclusively for unit tests. What do you think?

**Comment by blms** [comment]
Note for @rlskoeser: I had at first used a document with mirador for visual tests, but it broke on both Lighthouse and Percy. Not sure if we want to have one with mirador going forward, maybe it would be good for a separate issue to determine how to resolve all the Lighthouse errors and get it to display properly on Percy.

**Comment by rlskoeser** [comment]
> Note for @rlskoeser: I had at first used a document with mirador for visual tests, but it broke on both Lighthouse and Percy. Not sure if we want to have one with mirador going forward, maybe it would be good for a separate issue to determine how to resolve all the Lighthouse errors and get it to display properly on Percy.

Yes, when I was initially setting up lighthouse I fought with mirador for a while and finally decided to skip testing on it until we work on implementing local manifests and setting up the IIIF viewer the way we want it. Mirador is causing a ton of lighthouse errors and I'm not sure the right solution — not our job to fix them, but I'm not sure that telling lighthouse to ignore them is a useful approach either (seems like we'll ignore too many other things as well).

File: geniza/corpus/tests/test_corpus_models.py [hunk 1]
```diff
@@ -211,6 +213,24 @@ def test_str(self):
...
+        """Should use name as natural key"""
+        doc_type = DocumentType(name="SomeType")
+        assert len(doc_type.natural_key()) == 1
+        assert "SomeType" in doc_type.natural_key()
+        doc = Document.objects.create()
+        doc.doctype = doc_type
+        serialized_doc = serialize(
+            "json", [doc], use_natural_foreign_keys=True, use_natural_primary_keys=True
+        )
+        assert '"doctype": ["SomeType"]' in serialized_doc
```
Comment (rlskoeser) [inline geniza/corpus/tests/test_corpus_models.py:226]: this part seems unnecessary to me  — you're testing django logic, not local code

File: requirements/dev.txt [hunk 2]
```diff
@@ -7,5 +7,9 @@ pre-commit
 black
 #sphinx
 pytest-ordering
-django-fixture-magic
+# Using a fork of django-fixture-magic until this PR is merged:
+# https://github.com/davedash/django-fixture-magic/pull/76
+# (merge_fixtures with natural keys)
```
Comment (rlskoeser) [inline requirements/dev.txt:12]: thanks for the clear comment on the fork

**Review by rlskoeser — APPROVED** [review]
Updates look good. I'd prefer a more descriptive name for the fixture, but don't have any immediate suggestions. It's used for lighthouse and visual review, right? Maybe there's a brief name that covers both of them to distinguish from unit testing.

File: geniza/common/management/commands/create_test_site.py [hunk 3]
```diff
@@ -26,14 +19,8 @@ def handle(self, *args, **options):
...
-        # set up test models via pytest fixtures
-        frag = make_fragment()
-        multifrag = make_multifragment()
-        # remove the example iiif url used in the test fixture
-        multifrag.iiif_url = ""
-        multifrag.save()
-        make_document(multifrag)
-        make_join(frag, multifrag)
+        # set up test models via JSON fixtures
+        call_command("loaddata", "test_documents.json")
```
[the remark this answers is not in the document]
Comment (blms) [reply geniza/common/management/commands/create_test_site.py:23]: Maybe `ui_ux_test_documents.json`? (Thinking of Percy as UI and Lighthouse as UX, not sure if that makes sense)

[the remark this answers is not in the document]
Comment (rlskoeser) [reply geniza/common/management/commands/create_test_site.py:23]: @blms yes, I like that! it's short, and seems clear enough to distinguish from unit testing

**Pull request merged**